lib/promscrape/config.go

```diff
@@ -102,6 +102,52 @@ func (cfg *Config) mustStart() {
 	logger.Infof("started service discovery routines in %.3f seconds", time.Since(startTime).Seconds())
 }
 
+func (cfg *Config) mustStartRestart(prevCfg *Config) {
+	startTime := time.Now()
+	logger.Infof("restarting service discovery routines...")
+
+	prevScrapeCfgByName := make(map[string]*ScrapeConfig, len(prevCfg.ScrapeConfigs))
+	for i := range prevCfg.ScrapeConfigs {
+		sc := &prevCfg.ScrapeConfigs[i]
+		prevScrapeCfgByName[sc.JobName] = sc
+	}
+
+	var toStart, toStop []*ScrapeConfig
+	currentJobNames := make(map[string]struct{}, len(cfg.ScrapeConfigs))
+	for i := range cfg.ScrapeConfigs {
+		sc := cfg.ScrapeConfigs[i]
+		currentJobNames[sc.JobName] = struct{}{}
+		prevSC := prevScrapeCfgByName[sc.JobName]
+
+		if prevSC != nil && prevSC.checksum == sc.checksum {
+			// config checksum the same, no need to restart it
+			// replace reference with previous job, it allows to stop it properly later.
+			cfg.ScrapeConfigs[i] = *prevSC
+			continue
+		}
+		if prevSC != nil {
```

Comment:
```suggestion
		// prev scrape config isn't present in the current config, so we mark it to stop
		if prevSC != nil {
```

Reply:
It's a case, when prevConfig is present in current config, but it's checksum not match current checksum. I'll add proper comment.